File: internal/fileutil/fileutil.go
package fileutil

import (
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"sort"
	"strconv"
	"strings"
)

const Version = "v2.1.0"

var (
	Reset  = "\033[0m"
	Red    = "\033[31m"
	Green  = "\033[32m"
	Yellow = "\033[33m"
	Cyan   = "\033[36m"
	Bold   = "\033[1m"
	Dim    = "\033[2m"
)

type FileType int

const (
	TypeImage FileType = iota
	TypeVideo
	TypeAudio
	TypeUnknown
)

type FileInfo struct {
	Index    int
	Path     string
	Name     string
	Ext      string
	Size     int64
	SizeStr  string
	Type     FileType
	TypeName string
}

type AnalyseStats struct {
	Total     int
	TotalSize int64
	Images    int
	Videos    int
	Audio     int
	ImageSize int64
	VideoSize int64
	AudioSize int64
	Formats   map[string]int
	Dir       string
}

var (
	VideoExts = map[string]bool{
		".mp4": true, ".mov": true, ".avi": true, ".mkv": true,
		".wmv": true, ".flv": true, ".webm": true, ".m4v": true,
		".mpg": true, ".mpeg": true, ".3gp": true, ".ts": true,
	}
	ImageExts = map[string]bool{
		".jpg": true, ".jpeg": true, ".png": true, ".webp": true,
		".bmp": true, ".tiff": true, ".tif": true, ".avif": true,
		".gif": true, ".svg": true, ".ico": true, ".heic": true,
		".heif": true,
	}
	AudioExts = map[string]bool{
		".mp3": true, ".wav": true, ".flac": true, ".ogg": true,
		".aac": true, ".wma": true, ".m4a": true, ".opus": true,
		".aiff": true, ".alac": true,
	}
	FormatNames = map[string]string{
		".jpg": "JPEG", ".jpeg": "JPEG", ".png": "PNG", ".webp": "WebP",
		".avif": "AVIF", ".gif": "GIF", ".bmp": "BMP", ".svg": "SVG",
		".ico": "ICO", ".tiff": "TIFF", ".tif": "TIFF", ".heic": "HEIC",
		".heif": "HEIF",
		".mp4": "MP4", ".mov": "MOV", ".avi": "AVI", ".mkv": "MKV",
		".wmv": "WMV", ".flv": "FLV", ".webm": "WebM", ".m4v": "M4V",
		".mpg": "MPG", ".mpeg": "MPEG", ".3gp": "3GP", ".ts": "TS",
		".mp3": "MP3", ".wav": "WAV", ".flac": "FLAC", ".ogg": "OGG",
		".aac": "AAC", ".wma": "WMA", ".m4a": "M4A", ".opus": "Opus",
		".aiff": "AIFF", ".alac": "ALAC",
	}
)

func DetectType(ext string) FileType {
	if VideoExts[ext] {
		return TypeVideo
	}
	if ImageExts[ext] {
		return TypeImage
	}
	if AudioExts[ext] {
		return TypeAudio
	}
	return TypeUnknown
}

func CanConvertTo(ft FileType, format string) bool {
	f := strings.ToLower(format)
	switch ft {
	case TypeImage:
		switch f {
		case "webp", "avif", "jpg", "jpeg", "png", "gif", "bmp", "tiff":
			return true
		}
	case TypeVideo:
		switch f {
		case "mp4", "webm", "avi", "mov", "mkv", "gif":
			return true
		}
	case TypeAudio:
		switch f {
		case "mp3", "ogg", "wav", "flac", "aac", "opus", "m4a":
			return true
		}
	}
	return false
}

func FormatToType(format string) string {
	f := strings.ToLower(format)
	switch f {
	case "webp", "avif", "jpg", "jpeg", "png", "gif", "bmp", "tiff":
		return "image"
	case "mp4", "webm", "avi", "mov", "mkv":
		return "video"
	case "mp3", "ogg", "wav", "flac", "aac", "opus", "m4a":
		return "audio"
	}
	return "all"
}

func FormatSize(bytes int64) string {
	if bytes < 1024 {
		return strconv.FormatInt(bytes, 10) + " B"
	}
	if bytes < 1024*1024 {
		return strconv.FormatFloat(float64(bytes)/1024, 'f', 0, 64) + " KB"
	}
	if bytes < 1024*1024*1024 {
		return strconv.FormatFloat(float64(bytes)/1024/1024, 'f', 1, 64) + " MB"
	}
	return strconv.FormatFloat(float64(bytes)/1024/1024/1024, 'f', 2, 64) + " GB"
}

func FormatDisplayName(ext string) string {
	if n, ok := FormatNames[ext]; ok {
		return n
	}
	return strings.ToUpper(strings.TrimPrefix(ext, "."))
}

func FormatColor(ext string) string {
	if ImageExts[ext] {
		return Green
	}
	if VideoExts[ext] {
		return Cyan
	}
	if AudioExts[ext] {
		return Yellow
	}
	return Reset
}

func SortFiles(files []FileInfo) {
	sort.Slice(files, func(i, j int) bool {
		return files[i].Name < files[j].Name
	})
}

func FindFFmpeg() (string, string) {
	exePath, _ := os.Executable()
	exeDir := filepath.Dir(exePath)

	local := filepath.Join(exeDir, "ffmpeg.exe")
	if fileExists(local) {
		return local, exeDir
	}
	localBin := filepath.Join(exeDir, "bin", "ffmpeg.exe")
	if fileExists(localBin) {
		return localBin, exeDir
	}
	if path, err := exec.LookPath("ffmpeg"); err == nil {
		return path, exeDir
	}
	return "", exeDir
}

func FileExists(path string) bool {
	_, err := os.Stat(path)
	return err == nil
}

func FileNameWithoutExt(name string) string {
	ext := filepath.Ext(name)
	return name[:len(name)-len(ext)]
}

func IsValidTargetFormat(format string) bool {
	f := strings.ToLower(format)
	valid := map[string]bool{
		"webp": true, "avif": true, "jpg": true, "jpeg": true,
		"png": true, "gif": true, "bmp": true,
		"mp4": true, "webm": true, "avi": true, "mov": true, "mkv": true,
		"mp3": true, "ogg": true, "wav": true, "flac": true,
		"aac": true, "opus": true, "m4a": true,
	}
	return valid[f]
}

func FilterByType(files []FileInfo, filter string) []FileInfo {
	if filter == "all" || filter == "custom" {
		return files
	}
	var ft FileType
	switch filter {
	case "image":
		ft = TypeImage
	case "video":
		ft = TypeVideo
	case "audio":
		ft = TypeAudio
	default:
		return files
	}
	var result []FileInfo
	for _, f := range files {
		if f.Type == ft {
			result = append(result, f)
		}
	}
	return result
}

func FilterByFormat(files []FileInfo, format string) []FileInfo {
	var result []FileInfo
	for _, f := range files {
		if CanConvertTo(f.Type, format) {
			result = append(result, f)
		}
	}
	return result
}

func IsInPATH(dir string) bool {
	path := os.Getenv("PATH")
	separator := ";"
	if runtime.GOOS != "windows" {
		separator = ":"
	}
	for _, p := range strings.Split(path, separator) {
		if strings.EqualFold(strings.TrimSpace(p), dir) {
			return true
		}
	}
	return false
}

func fileExists(path string) bool {
	_, err := os.Stat(path)
	return err == nil
}

File: crates/crush-core/src/core/fileutil.rs
use std::path::Path;

#[derive(Debug, Clone, Copy, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
pub enum FileType {
    Image,
    Video,
    Audio,
    Unknown,
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct FileInfo {
    pub index: usize,
    pub path: String,
    pub name: String,
    pub ext: String,
    pub size: u64,
    pub size_str: String,
    pub file_type: FileType,
    pub type_name: String,
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct AnalyseStats {
    pub dir: String,
    pub total: usize,
    pub total_size: u64,
    pub images: usize,
    pub videos: usize,
    pub audio: usize,
    pub image_size: u64,
    pub video_size: u64,
    pub audio_size: u64,
    pub formats: std::collections::HashMap<String, usize>,
}

const VIDEO_EXTS: &[&str] = &[
    ".mp4", ".mov", ".avi", ".mkv", ".wmv", ".flv", ".webm", ".m4v", ".mpg", ".mpeg", ".3gp", ".ts",
];

const IMAGE_EXTS: &[&str] = &[
    ".jpg", ".jpeg", ".png", ".webp", ".bmp", ".tiff", ".tif", ".avif", ".gif", ".svg", ".ico",
    ".heic", ".heif",
];

const AUDIO_EXTS: &[&str] = &[
    ".mp3", ".wav", ".flac", ".ogg", ".aac", ".wma", ".m4a", ".opus", ".aiff", ".alac",
];

const FORMAT_NAMES: &[(&str, &str)] = &[
    (".jpg", "JPEG"),
    (".jpeg", "JPEG"),
    (".png", "PNG"),
    (".webp", "WebP"),
    (".avif", "AVIF"),
    (".gif", "GIF"),
    (".bmp", "BMP"),
    (".svg", "SVG"),
    (".ico", "ICO"),
    (".tiff", "TIFF"),
    (".heic", "HEIC"),
    (".mp4", "MP4"),
    (".mov", "MOV"),
    (".avi", "AVI"),
    (".mkv", "MKV"),
    (".wmv", "WMV"),
    (".flv", "FLV"),
    (".webm", "WebM"),
    (".m4v", "M4V"),
    (".mp3", "MP3"),
    (".wav", "WAV"),
    (".flac", "FLAC"),
    (".ogg", "OGG"),
    (".aac", "AAC"),
    (".m4a", "M4A"),
    (".opus", "Opus"),
];

pub fn detect_type(ext: &str) -> FileType {
    let lower = ext.to_lowercase();
    if VIDEO_EXTS.iter().any(|&e| e == lower) {
        FileType::Video
    } else if IMAGE_EXTS.iter().any(|&e| e == lower) {
        FileType::Image
    } else if AUDIO_EXTS.iter().any(|&e| e == lower) {
        FileType::Audio
    } else {
        FileType::Unknown
    }
}

pub fn format_display_name(ext: &str) -> String {
    let lower = ext.to_lowercase();
    for &(e, name) in FORMAT_NAMES {
        if e == lower {
            return name.to_string();
        }
    }
    ext.trim_start_matches('.').to_uppercase()
}

pub fn format_size(bytes: u64) -> String {
    if bytes < 1024 {
        return format!("{} B", bytes);
    }
    if bytes < 1024 * 1024 {
        return format!("{:.0} KB", bytes as f64 / 1024.0);
    }
    if bytes < 1024 * 1024 * 1024 {
        return format!("{:.1} MB", bytes as f64 / 1024.0 / 1024.0);
    }
    format!("{:.2} GB", bytes as f64 / 1024.0 / 1024.0 / 1024.0)
}

pub fn file_name_without_ext(name: &str) -> &str {
    Path::new(name)
        .file_stem()
        .and_then(|s| s.to_str())
        .unwrap_or(name)
}

pub fn can_convert_to(file_type: FileType, format: &str) -> bool {
    let f = format.to_lowercase();
    match file_type {
        FileType::Image => matches!(
            f.as_str(),
            "webp" | "avif" | "jpg" | "jpeg" | "png" | "gif" | "bmp" | "tiff"
        ),
        FileType::Video => matches!(
            f.as_str(),
            "mp4"
                | "webm"
                | "avi"
                | "mov"
                | "mkv"
                | "gif"
                | "mp3"
                | "ogg"
                | "wav"
                | "flac"
                | "aac"
                | "opus"
                | "m4a"
                | "alac"
        ),
        FileType::Audio => matches!(
            f.as_str(),
            "mp3" | "ogg" | "wav" | "flac" | "aac" | "opus" | "m4a" | "alac"
        ),
        FileType::Unknown => false,
    }
}

pub fn native_supports_format(input_ext: &str, target_ext: &str) -> bool {
    let inp = input_ext.to_lowercase();
    let tgt = target_ext.to_lowercase();

    let supported_input = matches!(
        inp.as_str(),
        ".png" | ".jpg" | ".jpeg" | ".bmp" | ".tiff" | ".tif" | ".gif"
    );
    let supported_output = matches!(
        tgt.as_str(),
        "webp" | "avif" | "png" | "jpg" | "jpeg" | "bmp"
    );

    supported_input && supported_output
}

pub fn scan_directory(dir: &Path) -> (Vec<FileInfo>, AnalyseStats) {
    let mut stats = AnalyseStats {
        dir: dir.display().to_string(),
        total: 0,
        total_size: 0,
        images: 0,
        videos: 0,
        audio: 0,
        image_size: 0,
        video_size: 0,
        audio_size: 0,
        formats: std::collections::HashMap::new(),
    };
    let mut files = Vec::new();

    let Ok(entries) = std::fs::read_dir(dir) else {
        return (files, stats);
    };

    let mut index = 0;
    for entry in entries.flatten() {
        let metadata = match entry.metadata() {
            Ok(m) => m,
            Err(_) => continue,
        };
        if metadata.is_dir() {
            continue;
        }

        let path = entry.path();
        let ext = path
            .extension()
            .and_then(|e| e.to_str())
            .unwrap_or("")
            .to_lowercase();

        if ext.is_empty() {
            continue;
        }

        let file_type = detect_type(&format!(".{}", ext));
        if file_type == FileType::Unknown {
            continue;
        }

        index += 1;
        let size = metadata.len();
        let name = entry.file_name().to_string_lossy().to_string();
        let type_name = format_display_name(&format!(".{}", ext));

        stats.total += 1;
        stats.total_size += size;
        *stats.formats.entry(ext.clone()).or_insert(0) += 1;

        match file_type {
            FileType::Image => {
                stats.images += 1;
                stats.image_size += size;
            }
            FileType::Video => {
                stats.videos += 1;
                stats.video_size += size;
            }
            FileType::Audio => {
                stats.audio += 1;
                stats.audio_size += size;
            }
            _ => {}
        }

        files.push(FileInfo {
            index,
            path: path.display().to_string(),
            name,
            ext,
            size,
            size_str: format_size(size),
            file_type,
            type_name,
        });
    }

    files.sort_by(|a, b| a.name.cmp(&b.name));
    for (i, f) in files.iter_mut().enumerate() {
        f.index = i + 1;
    }

    (files, stats)
}

pub fn filter_by_type(files: &[FileInfo], filter: &str) -> Vec<FileInfo> {
    let ft = match filter {
        "image" => FileType::Image,
        "video" => FileType::Video,
        "audio" => FileType::Audio,
        _ => return files.to_vec(),
    };
    files
        .iter()
        .filter(|f| f.file_type == ft)
        .cloned()
        .collect()
}
